code/controllers/subsystem/research.dm

```diff
@@ -86,28 +82,23 @@ SUBSYSTEM_DEF(research)
 	return SS_INIT_SUCCESS
 
 /datum/controller/subsystem/research/fire()
-	var/list/bitcoins = list()
-	for(var/obj/machinery/rnd/server/miner as anything in servers)
-		if(miner.working)
-			bitcoins = single_server_income.Copy()
-			break //Just need one to work.
-
-	if (!isnull(last_income))
-		var/income_time_difference = world.time - last_income
-		science_tech.last_bitcoins = bitcoins  // Doesn't take tick drift into account
-		for(var/i in bitcoins)
-			bitcoins[i] *= (income_time_difference / 10) * income_modifier
-		science_tech.add_point_list(bitcoins)
+	for(var/datum/techweb/techweb_list as anything in techwebs)
+		if(!techweb_list.should_generate_points)
+			continue
+		var/list/bitcoins = list()
+		for(var/obj/machinery/rnd/server/miner as anything in techweb_list.techweb_servers)
+			if(miner.working)
+				bitcoins = single_server_income.Copy()
+				break //Just need one to work.
 
-	last_income = world.time
+		if(!isnull(techweb_list.last_income))
```

Comment:
```suggestion
		if(isnull(techweb_list.last_income))
			continue
```

Reply:
This is how it was before, and we want it to be not null because we reference it to add points. The check is to make sure that the first tick doesn't generate points from a non-existent last income.